### PR TITLE
Attempt to clarify the signalling sequence in the spec

### LINKED ISSUE #202: cover the allowed call sequences on Subscriber in a spec rule
Currently it is only implicit that legal sequences start with `onError` or `onSubscribe` and end with `onComplete` or `onError` (or not at all), and it is not regulated within the spec rules that `onComplete` is illegal before `onSubscribe`.

I see two ways to approach this:

* we leave the protocol sequence definition outside of the spec and add the missing rule about `onComplete`
* we make the sequence definition explicit within the rules

The second approach would lead to duplication, since legal sequence grammars are almost perfectly constrained already, so I am leaning towards adding a new rule that orders `onComplete` and `onSubscribe`.

### PR DESCRIPTION
**Pull request by viktorklang**
@reactive-streams/contributors This is a proposal for spec amendment to fix #202 as discussed in that Issue, I will update the TCK appropriately if this proposal is accepted. Please vote before the 13th of February (or ask for an extension if you won't have time before that). Thank you!

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -57,8 +57,11 @@ A *Publisher* is a provider of a potentially unbounded number of sequenced eleme
 In response to a call to `Publisher.subscribe(Subscriber)` the possible invocation sequences for methods on the `Subscriber` are given by the following protocol:
 
 ```
-onError | (onSubscribe onNext* (onError | onComplete)?)
+onSubscribe onNext* (onError | onComplete)?
 ```
+
+This means that `onSubscribe` is always signalled,
+followed by a possibly unbounded number of `onNext` signals (as requested by `Subscriber`) followed by an `onError` signal if there is a failure, or an `onComplete` signal when no more elements are available.
```
Comment (drewhk) [inline README.md]: except when cancelled of course, then no onComplete or onError might arrive and onNext still being finitely called.

**Comment by drewhk** [comment]
LGTM apart from the nitpick

File: README.md (hunk 1 above)
Comment (viktorklang) [inline README.md]: That's a great catch, @drewhk, thank you!

File: README.md [hunk 2]
```diff
@@ -57,9 +57,12 @@ A *Publisher* is a provider of a potentially unbounded number of sequenced eleme
 In response to a call to `Publisher.subscribe(Subscriber)` the possible invocation sequences for methods on the `Subscriber` are given by the following protocol:
 
 ```
-onError | (onSubscribe onNext* (onError | onComplete)?)
+onSubscribe onNext* (onError | onComplete)?
 ```
 
+This means that `onSubscribe` is always signalled,
+followed by a possibly unbounded number of `onNext` signals (as requested by `Subscriber`) followed by an `onError` signal if there is a failure, or an `onComplete` signal when no more elements are available—all as long as the `Subscription` is not cancelled.
```
Comment (viktorklang) [inline README.md]: @drewhk WDYT about the updated text?

Comment (drewhk) [inline README.md]: Looks fine!

**Comment by benjchristensen** [comment]
I'm okay with these changes ... though still think it unnecessary to emit `onSubscribe` when just emitting a terminal event.

**Comment by viktorklang** [comment]
@benjchristensen Thanks for the input, Ben, I really appreciate it

File: README.md [hunk 3]
```diff
@@ -87,10 +90,10 @@ public interface Publisher<T> {
...
 | <a name="1.7">7</a>       | Once a terminal state has been signaled (`onError`, `onComplete`) it is REQUIRED that no further signals occur. |
 | <a name="1.8">8</a>       | If a `Subscription` is cancelled its `Subscriber` MUST eventually stop being signaled. |
-| <a name="1.9">9</a>       | Calling `Publisher.subscribe` MUST return normally except when the provided `Subscriber` is `null` in which case it MUST throw a `java.lang.NullPointerException` to the caller, for all other situations the only legal way to signal failure (or reject a `Subscriber`) is via the `onError` method. |
+| <a name="1.9">9</a>       | `Publisher.subscribe` MUST signal `onSubscribe` on the provided `Subscriber` prior to any other signals to that `Subscriber` and MUST return normally, except when the provided `Subscriber` is `null` in which case it MUST throw a `java.lang.NullPointerException` to the caller, for all other situations the only legal way to signal failure (or reject the `Subscriber`) is by calling `onError` (after calling `onSubscribe`). |
```
Comment (viktorklang) [inline README.md]: Hmmm, perhaps it should be: "`Publisher.subscribe` MUST call `onSubscribe`" for symmetry?

**Comment by viktorklang** [comment]
@reactive-streams/contributors I have now added appropriate TCK checks for the clarified signalling to this PR

**Comment by viktorklang** [comment]
Need to rebase this unto master, non-trivial so will hopefully get to it sometime tonight.

**Comment by viktorklang** [comment]
Phew, finally had time to get the rebase done. Merging after Travis OKs the branch

**Comment by viktorklang** [comment]
Noticed dangling reference to the "old" 1.12 that I repointed to 1.9, merging after Travis OKs

**Pull request merged**